File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        """
            1. init slow, fast ptr
            2. while slow.next and fast.next.next
            3. slow = slow.next, fast = fast.next.next
            4. if slow == fast:, return true
            5. return false
        """
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False
        

File: Data Structures & Algorithms/reverse-a-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) { return head; }

        ListNode* sub = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;

        return sub;
    }
};

/*
    base case: reverse a single node list. return node;
    recursive step: take a reversed list and reverse it and the element;
    1. do reassignment on head->next and reversedList(head->next);
*/
